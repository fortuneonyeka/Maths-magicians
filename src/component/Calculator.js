/* eslint-disable  */
import React from 'react';
import PropTypes from 'prop-types';

class Calculator extends React.Component {
  render() {
    const {
      handleClick, total, operation, next,
    } = this.props;
    return (

      <div className="container">
        <div className="display">
        { next || total ||operation|| '0' }
        </div>
        <div className="keypads">
          <button id="reset" name="AC" onClick={handleClick}>AC</button>
          <button id="backspace" name="+/-" onClick={handleClick}>+/-</button>
          <button name="%" onClick={handleClick}>%</button>
          <button className="operator" name="รท" onClick={handleClick}>รท</button>
          <button name="7" onClick={handleClick}>7</button>
          <button name="8" onClick={handleClick}>8</button>
          <button name="9" onClick={handleClick}>9</button>
          <button className="operator" name="x" onClick={handleClick}>x</button>
          <button name="4" onClick={handleClick}>4</button>
          <button name="5" onClick={handleClick}>5</button>
          <button name="6" onClick={handleClick}>6</button>
          <button className="operator" name="-" onClick={handleClick}>-</button>
          <button name="1" onClick={handleClick}>1</button>
          <button name="2" onClick={handleClick}>2</button>
          <button name="3" onClick={handleClick}>3</button>
          <button className="operator" id="plus" name="+" onClick={handleClick}>+</button>
          <button name="0" className="zero" onClick={handleClick}>0</button>
          <button name="." onClick={handleClick}>.</button>
          <button className="operator grid" onClick={handleClick}>=</button>
        </div>
      </div>

    );
  }
}

Calculator.PropTypes = {
  total: PropTypes.string.isRequired,
  operation: PropTypes.string.isRequired,
  next: PropTypes.string.isRequired,
  handleClick: PropTypes.func.isRequired
}

export default Calculator;