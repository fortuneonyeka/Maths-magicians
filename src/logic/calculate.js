import operate from './operate';

function isNumber (item) {
return !!item.match(/[0-9]+/);
}
