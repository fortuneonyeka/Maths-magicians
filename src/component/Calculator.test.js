import { render, fireEvent, screen } from '@testing-library/react';
// import { BrowserRouter as Router } from "react-router-dom";
import Calculator from './Calculator';
import calculate, { isNumber } from '../logic/calculate';

describe('Calculator Component', () => {
  it('Calculator snapshot test', () => {
    const { asFragment: calculator } = render(<Calculator />);
    expect(calculator()).toMatchSnapshot;
  });
});

describe("calculate", () => {
  it("IsNumber() should return false when input is not number", () => {
    expect(isNumber(".")).not.toBeTruthy();
  });

  it("IsNumber() should return true if the input is a number", () => {
    expect(isNumber("10")).toBeTruthy();
  });

  it("calculate({operation, next}, buttonName=AC) should return { total: null, next: null, operation: null }", () => {
    const buttonName = "AC";
    const obj = { total: 0, next: 0, operation: null };

    const result = calculate(obj, buttonName);

    expect(result).toEqual({ total:0, next: null, operation: null });
  });

  it("calculate({operation, next:0}, buttonName=0) should return {}", () => {
    const buttonName = "0";
    const obj = { total: null, next: "0", operation: null };

    const result = calculate(obj, buttonName);

    expect(result).toStrictEqual({});
  });

  it("calculate({operation, !next}, buttonName) should override obj.next: buttonName", () => {
    const buttonName = "1";
    const obj = { total: null, next: null, operation: "+" };

    const result = calculate(obj, buttonName);

    expect(result).toStrictEqual({ ...obj, next: buttonName });
  });

  it("calculate({operation, !next}, buttonName) should return obj.next + buttonName", () => {
    const buttonName = "4";
    const obj = { total: null, next: 10, operation: null };

    const result = calculate(obj, buttonName);

    expect(result).toStrictEqual({ total: null, next: 10 + buttonName });
  });
});


 describe("Calculator interaction tests", () => {
   let display;
   beforeEach(() => {
     const { container } = render(<Calculator />);
     display = container.querySelector(".display");
   });

   it("test interaction: 10 + 10 = 20, then press AC", () => {
     ["1", "0"].forEach((input) => {
       fireEvent.click(screen.getByText(input));
     });

     expect(display.innerHTML).toEqual("10");

     ["+", "1", "0", "="].forEach((input) => {
       fireEvent.click(screen.getByText(input));
     });

     expect(display.innerHTML).toEqual("20");

     fireEvent.click(screen.getByText("AC"));

     expect(display.innerHTML).toEqual("0");
   });

   it("test unacceptable interaction: + % =, should do nothing", () => {
     ["+", "%", "="].forEach((input) => {
       fireEvent.click(screen.getByText(input));
     });

     expect(display.innerHTML).toEqual("%");
   });

   it("test math error interaction: 1 % 0 prints `Can't divide by zero", () => {
     ["1", "%", "0", "="].forEach((input) => {
       fireEvent.click(screen.getByText(input));
     });

     expect(display.innerHTML).toEqual("Can't % 0.");
   });

   it("calculate({!operation, !next}, buttonName) should return next: buttonName", () => {
     const buttonName = "5";
     const obj = { total: null, next: null, operation: null };

     const result = calculate(obj, buttonName);

     expect(result).toStrictEqual({ next: buttonName, total: null });
   });

   
 });

 
