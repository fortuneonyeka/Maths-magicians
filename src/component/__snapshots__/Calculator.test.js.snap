// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component NavBar snapshot test 1`] = `
<DocumentFragment>
  <nav
    class="nav-menu"
  >
    <a
      class="nav-title"
      href="/"
    >
      Math Magicians
    </a>
    <ul
      class="nav-menu-list"
    >
      <li>
        <a
          activeclassname="active-link"
          aria-current="page"
          class="active"
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          activeclassname="active-link"
          class=""
          href="/calculator/"
        >
          Calculator
        </a>
      </li>
      <li>
        <a
          activeclassname="active-link"
          class=""
          href="/quote/"
        >
          Quote
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
